### PR TITLE
Implemented torch.sparse.max(sparse_tensor, dim)

### LINKED ISSUE #4247: Feature request: sparse matrix max(axis)
The current sparse support is pretty nice!  However, there are some important of collation operations that can not be efficiently be composed out of matmul or operations on _values().  One that is in the way of my current work is axis-wise max.  Ideally:

```
vec = sparsemat.max(axis=None, zero=0)
```

Where an optional zero argument could be used to specify the interpretation of unstored matrix entries.  (E.g., could pass -Infinity if that is the intent.)  In my application, I am using a sparse matrix to store only the significant log probabilities in a grid, and max() with a -Infinity default is the key operation needed to implement a stable row-wise logsumexp().)

For reference: scipy.sparse's inclusion of max is here, although they miss the need for a zero default.
https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.coo_matrix.max.html#scipy.sparse.csr_matrix.max

### PR DESCRIPTION
**Pull request by IAmKohlton**
Fixes #4247

This PR adds the function torch.sparse.max(sparse_tensor, dim). This was directly based off of torch.sparse.sum(sparse_tensor, dim). To do this I needed to implement a custom reduction function (which was mostly just a modified version of coalesce). It's important to note that this function isn't identical to the dense version of the function. The sparse implementation only takes the maximum over the non zero elements of the sparse tensor. This is similar to how sparse.log_softmax(sparse_tensor) operates already.

The function implemented in this PR does not support either CUDA or autograd. Those can come in future PRs

This is my first time contributing (cause I really need this function for a project I'm working on) so let me know if there's anything better I can do.

### CODE REVIEW

**Comment by ezyang** [comment]
@pearu or @IvanKobzarev, do you think either of you could look at this?

File: aten/src/ATen/native/sparse/SparseTensorMath.cpp [hunk 1]
```diff
@@ -1324,6 +1324,203 @@ Tensor _sparse_sum(const SparseTensor& input, IntArrayRef dims_to_sum) {
...
+  SparseTensor dst = at::native::new_with_dims_sparse(
+    sparse_dim,
+    dense_dim,
+    self.sizes(),
+    optTypeMetaToScalarType(self.options().dtype_opt()),
+    self.options().layout_opt(),
+    self.options().device_opt(),
+    self.options().pinned_memory_opt()
+  );
+  // TODO: is there a more idiomatic way to do this?
```
Comment (cpuhrsch) [inline aten/src/ATen/native/sparse/SparseTensorMath.cpp:1370]: You could use "empty_like(indices)"

File: aten/src/ATen/native/sparse/SparseTensorMath.cpp [hunk 2]
```diff
@@ -1324,6 +1324,203 @@ Tensor _sparse_sum(const SparseTensor& input, IntArrayRef dims_to_sum) {
...
+  std::tie(indicesBuffer, indicesPermutation) = indices_scalar.sort(0);
+  // NB: The accessor accesses here rely on self._nnz() > 0 (tested earlier in
+  // this function)
+  auto newIndicesAccessor = newIndices.accessor<int64_t, 2>();
+  auto indicesAccessor = indices.accessor<int64_t, 2>();
+  auto indicesPermutationAccessor = indicesPermutation.accessor<int64_t, 1>();
+  auto indicesBufferAccessor = indicesBuffer.accessor<int64_t, 1>();
+
+  int64_t i = -1;
+  AT_DISPATCH_ALL_TYPES(values.scalar_type(), "coalesce", [&] {
```
Comment (cpuhrsch) [inline aten/src/ATen/native/sparse/SparseTensorMath.cpp:1386]: Could you add a few more comments as to what this function does?

**Comment by IAmKohlton** [comment]
Quick update: I've been super busy with work, and moving apartments over the last month, but I get a week off work at the end of July. During that I'll do my best to clean it up and improve it

**Comment by cpuhrsch** [comment]
@IAmKohlton - as discussed in your other issue https://github.com/pytorch/pytorch/issues/61788 masked semantics are currently not well defined within PyTorch. Before I can accept expanded coverage for operators like hits, I'd want to first settle on the semantics of masked reductions (at a minimum). I'm currently writing up an RFC on this matter.

**Comment by IAmKohlton** [comment]
@cpuhrsch Makes perfect sense, I completely understand. In the mean time if I modified it to be mathematically consistent to the dense max, would you reconsider it?

**Comment by cpuhrsch** [comment]
@IAmKohlton, I think we could even go as far as simply adding support for Sparse to torch.max. Let me check in with my team and get back to you.

**Comment by cpuhrsch** [comment]
@IAmKohlton - we'll need more time to distill a decision here. For now we won't be able to move forward with this PR.

**Comment by Doublelucker** [comment]
And here I am needing exactly this function. When is this going to be resolved?

**Comment by cpuhrsch** [comment]
We just landed the [masked amax function](https://github.com/pytorch/pytorch/pull/65990). Next steps include support for sparse layouts, but I don't want to set false expectations for a timeline here.

**Comment by IvanYashchuk** [comment]
It might be worth to revisit this PR.

**Pull request closed without merging**